### ISSUE TITLE
warning on foreach conversions

### ISSUE THREAD

**Issue by ZacharyPatten** [dotnet/roslyn-analyzers#5672]
> this was originally posted in roslyn, and moved here: https://github.com/dotnet/roslyn/discussions/57308

There should be a compiler warning when a `foreach` loop compiles and adds an `explicit` cast operation. Here is an example that would apply:
```cs
// Example #1: foreach adds explicits casts between enum types
using System;

foreach (A a in Enum.GetValues<B>()) // <- warning
{
	Console.WriteLine(a);
}

enum A { A = 'a', B = 'b', C = 'c', }
enum B { D = 'd', E = 'e', F = 'f', }

// Output:
// 100
// 101
// 102
```

Another example:
```cs
// Example #2: foreach adds explicits casts long -> int
long[] array = { long.MinValue, 1, 2, 3, long.MaxValue };
foreach (int i in array) // <- warning
{
	Console.WriteLine(i);
}
// Output:
// 0
// 1
// 2
// 3
// -1
```

From my experience, this behavior often results in bugs in code. I honeslty prefer this code not compile, but if there was at least a warning then I can just increase the severity of that warning to be an error and thus break builds.

## Acknowledging This May Be An Annoying Warning

If this warning was added there currently isn't any clean syntax to explicitly allow explicit casts in `foreach` loops. So develpoers who cannot simply ignore or suppress it may find this warning annoying. Here are two examples that modify `Example #2` so that it would not result in the proposed warning:

```cs
// Example #3: modify Example #2 to bypass warning via casting in loop
long[] array = { long.MinValue, 1, 2, 3, long.MaxValue };
foreach (long i in array)
{
	int i2 = (int)i;
	Console.WriteLine(i2);
}
```

```cs
// Example #4: modify Example #2 to bypass warning via Linq Select
long[] array = { long.MinValue, 1, 2, 3, long.MaxValue };
foreach (int i in array.Select(l => (int)l))
{
	Console.WriteLine(i);
}
```

So while there currently would be ways to bypass the warning via syntax (rather than suppressing/ignoring it), those ways are not really ideal.

## More Good Than Harm

In my opinion, adding this compiler warning would do more good than harm. I would assume intentionally allowing explicit casts from `foreach` loops is relatively uncommon and accidentally using the wrong type in a foreach loop is relatively common (especially for new developers). Without a compiler warning these bugs can sometimes be hard to find. With a compiler warning (that we can escalate to an error if we choose), we can prevent these bugs with ease.

## What About Implicit Casts?

It may be nice if there was a warning for implicit casts too (not just explicit casts as mentioned in this post). If that was considered, it would be nice if it was a seperate warning so we can configure implicit and explicit casts seperately. Also, for implicit casts, maybe that would be better as a compiler message than a warning since it would seem less likely to be a bug in the code.

**Comment by mavasani** [comment]
Duplicate of https://github.com/dotnet/runtime/issues/48529, which was not approved by .NET runtime team due to concern over far too many false positives.

**Comment by RikkiGibson** [comment]
I think it's worth investigating whether there's some heuristics which could reduce "noise" warnings here to an acceptable level. For example, if the type of `Current` on the enumerator is `object`, then let's not give a diagnostic on the implicit cast. This would rule out usability issues around non-generic collections and likely a lot of other scenarios where people really do need to downcast via the variable type.

I would love to see data on how many "noise" diagnostics we get when building roslyn or the runtime with this. It feels like if you are enumerating something that doesn't return `object`s yet you depend on the implicit casting via the loop variable, what are you doing? I genuinely would like to know. I would not be surprised if we uncover some bugs in the process.

**Comment by RikkiGibson** [comment]
Closing as a duplicate of dotnet/roslyn#54712.
